src/lib.rs

```diff
@@ -145,10 +145,22 @@ fn to_avro_value(py: Python, datum: &PyObject, schema: &Schema) -> PyResult<Valu
 
             Ok(Value::Map(items))
         }
-        &Schema::Union(_) if datum.is_none() => Ok(Value::Union(None)),
-        &Schema::Union(ref inner) => Ok(Value::Union(Some(Box::new(to_avro_value(
-            py, datum, inner,
-        )?)))),
+        &Schema::Union(ref inner) => {
+            // Optimization for when union is used for optional values
+            if inner.is_nullable() && datum.is_none() {
+                Ok(Value::Union(Box::new(Value::Null)))
+            } else {
+                let variants = inner.variants();
+                for variant in variants {
```

Comment:
This is also the approach used in the upstream avro repository, so I don't think you are doing anything wrong here 🙂

https://github.com/apache/avro/blob/544e372c63daa720a01b4d558d53fb12aa6016b8/lang/py/src/avro/io.py#L1015